### ISSUE TITLE
How to create folder?

### ISSUE THREAD

**Issue by allspace** [minio/minio#2423]
I see comments like below:

// Minio does not support object names with trailing "/".

But AWS S3 uses trailing slash to create folder. So, How can I create folder with minio?

**Comment by krishnasrinivas** [comment]
@allspace what AWS S3 does is it creates an object with trailing "/" but the contents of the object is of 0 bytes to simulate a directory.

For minio we decided to use simple file system (ext4, xfs etc) as backend without using any database to store objects. Hence an object is just a file in the backend and if there is a "/" in the object name we create a directory in the backend. Hence we can't create objects with trailing "/"

**Comment by balamurugana** [comment]
@allspace To add up more, there is no empty directories in minio. Whenever an object is created with directory prefix, the prefix directory is created along with object automatically e.g. When `photos/sfo.jpg` object is created `photos` directory is also created and upon removal of `photos/sfo.jpg`, `photos` directory is deleted if its empty.

**Comment by allspace** [comment]
I'd suggest you add support for empty directory.

This is to not only be consistent with AWS S3 behavior, but also be consistent with minio's other APIs, e.g. ListObject can list an empty directory just as AWS S3 does.

**Comment by balamurugana** [comment]
@allspace Sure. This feature can be added if its absolutely necessary.  Currently its not found as compatibility issue.

**Comment by StephanX** [comment]
It'd be super helpful in my case.

I'm looking to use minio in kubernetes as both an S3 object store, and a shared fileystem.  Not being able to create a directory object with s3fs breaks any sort of recursive copy operation.

```
root@myminiotest-0:/# mkdir -p my/fancy/directory
root@myminiotest-0:/# echo "some data" > some/fancy/directory/data.txt
root@myminiotest-0:/# s3fs myfancybucket /mybucket -o use_path_request_style,url=http://100.69.131.19:9000
root@myminiotest-0:/# cp -a some /mybucket
root@myminiotest-0:/# cp: cannot create directory '/mybucket/some': No such file or directory
```
If there's a storagecontroller in kubernetes analogous to https://kubernetes.io/docs/concepts/storage/volumes/#cephfs that would work just fine as well.

Thanks!

**Comment by harshavardhana** [comment]
Possible fix available in the proposal - https://github.com/minio/minio/pull/4426

**Comment by harshavardhana** [comment]
> Possible fix available in the proposal - #4426

A major problem which even the #4426 will not be able to fix is s3fs setting odd metadata on the directories

```
map[content-type:application/x-directory X-Amz-Meta-Mode:493 X-Amz-Meta-Mtime:1498614550 X-Amz-Meta-Gid:0 X-Amz-Meta-Uid:0 etag:]
```

But this can be worked around by just using `complement_stat` mount option only available in the latest release 1.82 - https://github.com/s3fs-fuse/s3fs-fuse/commit/758b92e823540c73032412423bee6d8c93584148

**Comment by harshavardhana** [comment]
> A major problem which even the #4426 will not be able to fix is s3fs setting odd metadata on the directories
> 
> map[content-type:application/x-directory X-Amz-Meta-Mode:493 X-Amz-Meta-Mtime:1498614550 X-Amz-Meta-Gid:0 X-Amz-Meta-Uid:0 etag:]
> But this can be worked around by just using complement_stat mount option only available in the latest release 1.82 - s3fs-fuse/s3fs-fuse@758b92e

After much testing PR #4426 + minio master  and s3fs without any modification and no more special options. It seems to be working fine.. 

```
$ cp -r /usr/sbin .
/usr/sbin/zic:             31.32 MB / 31.32 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 1.61 MB/s 19s$ ls
[2017-06-27 20:53:46 PDT] 4.0KiB sbin/
$ ls sbin/

$ df . -h
Filesystem      Size  Used Avail Use% Mounted on
s3fs            256T     0  256T   0% /mnt/minfs

$ mount |grep minfs
s3fs on /mnt/minfs type fuse.s3fs (rw,nosuid,nodev,relatime,user_id=0,group_id=0)

$ s3fs -f harshavardhana: /mnt/minfs -o use_path_request_style,url=http://10.0.0.13:9000

```

One caveat in #4426 is it still doesn't support directories on erasure coded backend. I tried doing that for a bit but then i realized it can get complicated so it has been left as is.

**Comment by harshavardhana** [comment]
https://github.com/minio/minio/pull/5049 - FS now has support for empty directories so s3fs works fine for all operations now (including directory operations) - continuing from the comment above @allspace.  

Erasure coded backend is not part of this change and there are no current plans on addressing this for erasure coded backend. We can close this bug @deekoder as fixed.

**Comment by ptemmer** [comment]
Pardon my ignorance, from what I read here creating an empty directory should work.
However, what I am seeing is:

```
/data # mkdir test
mkdir: can't create directory 'test': No such file or directory
/data # df . -h
Filesystem                Size      Used Available Use% Mounted on
s3fs                    256.0T         0    256.0T   0% /data
/data # mount | grep s3fs
s3fs on /data type fuse.s3fs (rw,nosuid,nodev,relatime,user_id=0,group_id=0)
```
I'm using the latest minio docker image (created 13 days ago), with s3fs 1.79

**Comment by harshavardhana** [comment]
> Pardon my ignorance, from what I read here creating an empty directory should work.
> However, what I am seeing is:
> 
> /data # mkdir test
> mkdir: can't create directory 'test': No such file or directory
> /data # df . -h
> Filesystem                Size      Used Available Use% Mounted on
> s3fs                    256.0T         0    256.0T   0% /data
> /data # mount | grep s3fs
> s3fs on /data type fuse.s3fs (rw,nosuid,nodev,relatime,user_id=0,group_id=0)
> I'm using the latest minio docker image (created 13 days ago), with s3fs 1.79

Are you using distributed or single disk mode? @ptemmer - if former is true we recently merged a PR to support it.  For single disk mode it should work fine.

**Comment by ptemmer** [comment]
Thanks for the prompt reply @harshavardhana. Yes I am using distributed mode (4 nodes, 1 disk).
Is the merge already available in edge? When do you foresee it to be available in stable?
I would be glad to test.

**Comment by harshavardhana** [comment]
> Thanks for the prompt reply @harshavardhana. Yes I am using distributed mode (4 nodes, 1 disk).
> Is the merge already available in edge? When do you foresee it to be available in stable?
> I would be glad to test.

Yes it is in edge but i found a bug so this PR https://github.com/minio/minio/pull/5480 needs to be merged before you can test it.

**Comment by harshavardhana** [comment]
> Thanks for the prompt reply @harshavardhana. Yes I am using distributed mode (4 nodes, 1 disk).
> Is the merge already available in edge? When do you foresee it to be available in stable?
> I would be glad to test.

https://github.com/minio/minio/pull/5480 PR is merged @ptemmer you can test `edge` now which should work fine with s3fs.

**Comment by wimgoeman** [comment]
Heads up for anyone stumbling upon this:

If you want to create a folder, by creating an object with trailing '/', make sure its **size is 0 bytes**.  If its not 0 bytes, you will get back an error that the object name is invalid even though it is just the size which is the issue.
